### PR TITLE
DAOS-15748 dfuse: Have pre read work with out-of-order I/O

### PR DESCRIPTION
**Pull request by ashleypittman**
Handle out-of-order reads in the pre-read code and improve the test.
The code was returning correct data but wasn't using the new feature
if I/O had been re-ordered, change the enable logic to detect I/O that
might fall into this category and use the feature for well sized but
dis-ordered I/O.  Update the linear-read check in this case so that
it's still enabled for future files.

Improve the test by checking a second file after the first to ensure
the feature is left enabled.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/tests/ftest/dfuse/read.py
Comment (daltonbohning) [inline src/tests/ftest/dfuse/read.py:23]:
```suggestion
        the first file leaves the flag enabled.
```

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

**Review by jolivier23 — APPROVED** [review]

**Review by johannlombardi — APPROVED** [review]

**Comment by mchaarawi** [comment]
why did you restart testing on this? 
why are you pushing in the queue as high priority?

**Review by mchaarawi — CHANGES_REQUESTED** [review]
blocking landing because of failed dfuse tests with `Features: dfuse` here
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-14248/6/tests
and master merge commit does not have that features tag. sorry.
also this ticket is not marked as required for 2.6 so you should not rebuild with P2.

**Comment by ashleypittman** [comment]
> blocking landing because of failed dfuse tests with `Features: dfuse` here https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-14248/6/tests and master merge commit does not have that features tag.

This is a bug-fix for one of three dfuse tests which are currently failing.  This is one is an intermittent failure, https://github.com/daos-stack/daos/pull/14262 is for something which is always failing and https://github.com/daos-stack/daos/pull/14256 is intermittently failing for everyone, even without the dfuse tag.  Until we can get these three landed then `Features: dfuse` will be not pass and is causing wasted CI cycles.

**Comment by mchaarawi** [comment]
> > blocking landing because of failed dfuse tests with `Features: dfuse` here https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-14248/6/tests and master merge commit does not have that features tag.
> 
> This is a bug-fix for one of three dfuse tests which are currently failing. This is one is an intermittent failure, #14262 is for something which is always failing and #14256 is intermittently failing for everyone, even without the dfuse tag. Until we can get these three landed then `Features: dfuse` will be not pass and is causing wasted CI cycles.

i disagree. it is required unless all extra tests with the Features tag are failing. just because some tests are failing does not mean you should not be using the features tag.

**Comment by mchaarawi** [comment]
> > blocking landing because of failed dfuse tests with `Features: dfuse` here https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-14248/6/tests and master merge commit does not have that features tag.
> 
> This is a bug-fix for one of three dfuse tests which are currently failing. This is one is an intermittent failure, #14262 is for something which is always failing and #14256 is intermittently failing for everyone, even without the dfuse tag. Until we can get these three landed then `Features: dfuse` will be not pass and is causing wasted CI cycles.

and again https://daosio.atlassian.net/browse/DAOS-15748 is not required for 2.6 (and it really shouldn't block 2.6). so please respect the order of the testing queue like everyone else.

**Comment by mchaarawi** [comment]
> This is a bug-fix for one of three dfuse tests which are currently failing. This is one is an intermittent failure, #14262 is for something which is always failing and #14256 is intermittently failing for everyone, even without the dfuse tag. Until we can get these three landed then `Features: dfuse` will be not pass and is causing wasted CI cycles.

afaik there is also a way in the commit pragma to exclude running a test with a particular tag if that is known to be failing (using the minus sign i believe).

**Comment by ashleypittman** [comment]
Failures look to be DAOS-15124.  Failures in previous build which ran with `Features: dfuse` are DAOS-15662, DAOS-15745, DAOS-15616 and DAOS-15747

**Comment by ashleypittman** [comment]
> > afaik there is also a way in the commit pragma to exclude running a test with a particular tag if that is known to be failing (using the minus sign i believe).
> 
> That would be
> 
> ```
> Features: dfuse,-test_that_is_failing
> ```

I'll make a note of that, thank you.

**Comment by mchaarawi** [comment]
> Failures look to be DAOS-15124. Failures in previous build which ran with `Features: dfuse` are DAOS-15662, DAOS-15745, DAOS-15616 and DAOS-15747

this is a lot of failures.. you stopped the previous build before it has completed, so it did not run fully and we do not know if there are other issues. As for the issue you linked:
15662 is resolved and should pass with rebase.
15748 is resolved
15745 is close to land
15747 is resolved

so that leaves 15616. I would suggest you wait till 15745 is landed, then rebase and rerun with the Feature tag excluding only the test from 15616.
removing gatekeeper request as my concern with testing is not addressed.

**Pull request merged**
